packages/emotion-hash/hash.ml

```diff
@@ -1,133 +1,72 @@
+  (* 'm' and 'r' are mixing constants generated offline.
+     They're not really 'magic', they just happen to work well. *)
+  let m = 0x5bd1e995l in
+  let r = 24 in
+
+  (* Initialize the hash *)
+  let h = ref 0l in
 
-let to_css (number : Int32.t) = number |> to_base36
+  (* Mix 4 bytes at a time into the hash *)
+  let i = ref 0 in
+  let len = ref (String.length str) in
+  let of_char c = of_int (int_of_char c) in
+  while !len >= 4 do
+    let k =
+      of_char str.[!i]
+      land 0xffl
+      lor ((of_char str.[!i + 1] land 0xffl) lsl 8)
+      lor ((of_char str.[!i + 2] land 0xffl) lsl 16)
+      lor ((of_char str.[!i + 3] land 0xffl) lsl 24)
+    in
 
-(*
-    The murmur2 hashing is based on @emotion/hash, which is based on
-    https://github.com/garycourt/murmurhash-js and ported from
-    https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash2.cpp#L37-L86.
-    Reference: https://github.com/emotion-js/emotion/blob/main/packages/hash/src/index.js
+    let k = add (mul (k land 0xffffl) m) (mul (k lsr 16) 0xe995l lsl 16) in
+    let k = k lxor (k lsr r) in
 
-    It's an ongoing effort to have a hashing function, currently okish. *)
-let murmur2 (str : string) =
-  let length = String.length str in
-  (* var h = 0 *)
-  let hash = ref 0l in
-  (* var k *)
-  let k = ref 0l in
-  (* len = str.length *)
-  let len = ref length in
-  (* i = 0, *)
-  let index = ref 0 in
-  while !index + 4 <= length do
-    let char_code_1 = Int32.of_int (Char.code str.[!index]) in
-    let char_code_2 = Int32.of_int (Char.code str.[!index + 1]) in
-    let char_code_3 = Int32.of_int (Char.code str.[!index + 2]) in
-    let char_code_4 = Int32.of_int (Char.code str.[!index + 3]) in
-    (* k =
-       (str.charCodeAt(i) & 0xff) |
-       ((str.charCodeAt(++i) & 0xff) << 8) |
-       ((str.charCodeAt(++i) & 0xff) << 16) |
-       ((str.charCodeAt(++i) & 0xff) << 24) *)
-    k :=
-      char_code_1
-      ||| (char_code_2 << 8)
-      ||| (char_code_3 << 16)
-      ||| (char_code_4 << 24);
-    (* k =
-       (k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16) *)
-    k :=
-      ((!k & 0xffffl) * 0x5bd1e995l) ++ (((!k >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16;
-    (* k ^=  k >>> 24 *)
-    k := !k ^ (!k >>> 24);
-    k :=
-      ((!k & 0xffffl) * 0x5bd1e995l) ++ (((!k >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16;
-    (* h =
-       ((k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16)) ^
-       ((h & 0xffff) * 0x5bd1e995 + (((h >>> 16) * 0xe995) << 16)) *)
-    hash :=
-      (((!hash & 0xffffl) * 0x5bd1e995l)
-       ++ (((!hash >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16)
-      ^ !k;
-    index := !index + 1;
+    h :=
+      add (mul (k land 0xffffl) m) (mul (k lsr 16) 0xe995l lsl 16)
+      lxor add (mul (!h land 0xffffl) m) (mul (!h lsr 16) 0xe995l lsl 16);
+
+    i := !i + 4;
     len := !len - 4
   done;
 
   (* Handle the last few bytes of the input array *)
-  (match !len with
```

Comment:
This was problematic. In the original implementation, there's a `switch` with 3 cases, but there's no `break` on each case, so if `len` is 3, all the cases will be executed.

Using `match` doesn't follow that behavior, as only one branch is executed, it'd be analog to JS code having `break` on each case.

I ended up doing a sequence of `if` statements without `else`s, see below.